File: ThreadPool/ThreadPool/MyTask.cs
using System;
using System.Collections.Generic;
using System.Threading;

namespace ThreadPool
{
    public class MyTask<TResult> : IMyTask<TResult>
    {
        private readonly Func<TResult> _func;
        private volatile bool _isCompleted;
        private readonly object _lock = new();
        private readonly List<Exception> _exceptions = new();
        private TResult _result;

        public MyTask(Func<TResult> func)
        {
            _func = func ?? throw new ArgumentNullException(nameof(func), "Task cannot execute null");
        }

        public bool IsCompleted => _isCompleted;

        public TResult Result
        {
            get
            {
                Execute();
                if (_exceptions.Count > 0)
                {
                    throw new AggregateException(_exceptions);
                }

                return _result;
            }
        }

        public void Execute()
        {
            lock (_lock)
            {
                if (_isCompleted) return;
                try
                {
                    _result = _func.Invoke();
                }
                catch (AggregateException aggEx)
                {
                    _exceptions.AddRange(aggEx.InnerExceptions);
                }
                catch (Exception e)
                {
                    _exceptions.Add(e);
                }
                finally
                {
                    _isCompleted = true;
                }
            }
        }

        public IMyTask<TNewResult> ContinueWith<TNewResult>(Func<TResult, TNewResult> func)
        {
            return new MyTask<TNewResult>(() => func(Result));
        }
    }
}

File: task2/src/main/kotlin/spbu/mm/parallel/programming/MyTask.kt
package spbu.mm.parallel.programming

import kotlin.collections.ArrayList

class AggregateException(val exceptions: List<Exception>, message: String? = exceptions[0].message) : Exception(message, exceptions[0])

class MyTask<TResult>(val func: () -> TResult) : IMyTask<TResult> {
    @Volatile
    private var _isCompleted = false
    override val isCompleted: Boolean
        get() = _isCompleted

    private var _result: TResult? = null
    override val result: TResult
        get() {
            execute()
            if (exceptions.isNotEmpty()) throw AggregateException(exceptions)
            _isCompleted = true
            return _result!!
        }

    private val lock = object {}
    private val exceptions: ArrayList<Exception> = ArrayList()

    override fun <TNewResult> continueWith(func: (TResult) -> TNewResult): IMyTask<TNewResult> {
        return MyTask { func(result) }
    }

    override fun execute() {
        synchronized(lock) {
            if (_isCompleted) return
            try {
                _result = func()
            } catch (ex: AggregateException) {
                exceptions.addAll(ex.exceptions)
            } catch (ex: Exception) {
                exceptions.add(ex)
            } finally {
                _isCompleted = true
            }
        }
    }
}
